File: reverse-only-letters/Solution917.java
// https://leetcode.com/problems/reverse-only-letters/

import java.util.Stack;

class Solution917 {

    // T: O(n), S: O(n)
    public String reverseOnlyLetters(String s) {
        int left = 0, right = s.length() - 1;
        var arr = s.toCharArray();
        while (left < right) {
            char leftChar = arr[left];
            char rightChar = arr[right];
            if (Character.isLetter(leftChar) && Character.isLetter(rightChar)) {
                swap(arr, left, right);
                left++;
                right--;
            } else if (!Character.isLetter(leftChar)) {
                left++;
            } else if (!Character.isLetter(rightChar)) {
                right--;
            }
        }
        return new String(arr);
    }

    // T: O(n), S: O(n)
    public String reverseOnlyStringStack(String s) {
        var st = new Stack<Character>();
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (Character.isLetter(ch)) {
                st.push(ch);
            }
        }
        var sb = new StringBuilder();
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (Character.isLetter(ch)) {
                sb.append(st.pop());
            } else {
                sb.append(ch);
            }
        }
        return sb.toString();
    }

    private void swap(char[] arr, int i, int j) {
        if (i != j) {
            char temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }

    public static void main(String[] args) {
        var sol = new Solution917();
        var s1 = "ab-cd";
        var s2 = "a-bC-dEf-ghIj";
        var s3 = "Test1ng-Leet=code-Q!";
        System.out.println(sol.reverseOnlyLetters(s1)); // "dc-ba"
        System.out.println(sol.reverseOnlyStringStack(s1)); // "dc-ba"
        System.out.println(sol.reverseOnlyLetters(s2)); // "j-Ih-gfE-dCba"
        System.out.println(sol.reverseOnlyStringStack(s2)); // "j-Ih-gfE-dCba"
        System.out.println(sol.reverseOnlyLetters(s3)); // "Qedo1ct-eeLg=ntse-T!"
        System.out.println(sol.reverseOnlyStringStack(s3)); // "Qedo1ct-eeLg=ntse-T!"
    }

}

File: reverse-only-letters/solution917.go
// https://leetcode.com/problems/reverse-only-letters/

package main

import (
	"fmt"
	"strings"
	"unicode"
)

// T: O(n), S: O(n)
func reverseOnlyLetters(s string) string {
	runes := []rune(s)
	left, right := 0, len(s)-1
	for left < right {
		leftChar, rightChar := runes[left], runes[right]
		if unicode.IsLetter(leftChar) && unicode.IsLetter(rightChar) {
			runes[left], runes[right] = runes[right], runes[left]
			left++
			right--
		} else if !unicode.IsLetter(leftChar) {
			left++
		} else if !unicode.IsLetter(rightChar) {
			right--
		}
	}
	return string(runes)
}

// T: O(n), S: O(n)
func reverseOnlyLettersStack(s string) string {
	st := make([]rune, 0)
	for _, ch := range s {
		if unicode.IsLetter(ch) {
			st = append(st, ch)
		}
	}
	var sb strings.Builder
	for _, ch := range s {
		if unicode.IsLetter(ch) {
			lastIdx := len(st) - 1
			sb.WriteRune(st[lastIdx])
			st = st[:lastIdx]
		} else {
			sb.WriteRune(ch)
		}
	}
	return sb.String()
}

func main() {
	s1 := "ab-cd"
	s2 := "a-bC-dEf-ghIj"
	s3 := "Test1ng-Leet=code-Q!"
	fmt.Println(reverseOnlyLetters(s1))      // "dc-ba"
	fmt.Println(reverseOnlyLettersStack(s1)) // "dc-ba"
	fmt.Println(reverseOnlyLetters(s2))      // "j-Ih-gfE-dCba"
	fmt.Println(reverseOnlyLettersStack(s2)) // "j-Ih-gfE-dCba"
	fmt.Println(reverseOnlyLetters(s3))      // "Qedo1ct-eeLg=ntse-T!"
	fmt.Println(reverseOnlyLettersStack(s3)) // "Qedo1ct-eeLg=ntse-T!"
}
